r2-core/src/main/java/com/linkedin/r2/message/rest/RestRequestImpl.java

```diff
@@ -33,10 +33,22 @@
   /* package private */ RestRequestImpl(ByteString entity, Map<String, String> headers,
                                         List<String> cookies, URI uri, String method)
   {
-    super(headers, cookies, uri, method);
+    this(entity, headers, cookies, uri, method, true);
+  }
+
+  private RestRequestImpl(ByteString entity, Map<String, String> headers,
+      List<String> cookies, URI uri, String method, boolean validateArgs)
+  {
+    super(headers, cookies, uri, method, validateArgs);
     _entity = entity;
   }
 
+  @Override
+  public RestRequest copyWithURI(URI uri)
```

Comment:
This feels weird to me. Given a `RestRequest`, you want to provide a convenient method to get another RestRequest with only URI changed, this can be done completely outside pegasus, because `RestRequest` already provide a `builder` for it. So client should be very easy to do this by

```
   RestRequest origRequest = ....
   RestRequest updatedRequest = origRequest.builder().setURI(uri).build();
```
If this is the only purpose of this PR, I don't think that we need this here. Because  by analogy, I may question why we don't add any other convenient methods to update some part of an existing RestRequest. Personally feel current way to expose a `builder` API is cleaner.

Reply:
That is exactly what the default method in the interface does. However, calling this method can result in onerous performance costs, especially if it is in the hot path, since creating a builder and then building a request from it, involves some header copying and name validation.

I sent you a flamegraph offline showing the CPU impact of this on one of our internal services.